data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -36,118 +37,126 @@ public class ScanObjectWorker implements Runnable{
-        bucketOwnerProvider.getBucketOwner(scanOptions.getBucket())
-                .ifPresent(listObjectsV2Request::expectedBucketOwner);
-
-        if(Objects.nonNull(s3ScanKeyPathOption) && Objects.nonNull(s3ScanKeyPathOption.getS3ScanExcludeSuffixOptions()))
-            excludeItems.addAll(s3ScanKeyPathOption.getS3ScanExcludeSuffixOptions());
-
-        if(Objects.nonNull(s3ScanKeyPathOption) && Objects.nonNull(s3ScanKeyPathOption.getS3scanIncludeOptions()))
-            s3ScanKeyPathOption.getS3scanIncludeOptions().forEach(includePath -> {
-                listObjectsV2Request.prefix(includePath);
-                scanObjects.addAll(listS3Objects(excludeItems, listObjectsV2Request));
-            });
-        else
-            scanObjects.addAll(listS3Objects(excludeItems, listObjectsV2Request));
-
-        if(scanObjects.isEmpty())
-            LOG.info("s3 objects are not found in configured scan pipeline.");
-
-        scanObjects.forEach(key ->
-                processS3ObjectKeys(S3ObjectReference.bucketAndKey(scanOptions.getBucket(),
-                        key).build(),s3ObjectHandler, scanOptions));
+    // For testing
+    void runWithoutInfiniteLoop() {
+        startProcessingObject(10);
     }
 
-    private List<String> listS3Objects(List<String> excludeKeyPaths, ListObjectsV2Request.Builder listObjectsV2Request) {
-          return s3Client.listObjectsV2(listObjectsV2Request.fetchOwner(true).build()).contents().stream().map(S3Object::key)
-                .filter(path -> !path.endsWith("/"))
-                .filter(includeKeyPath -> excludeKeyPaths.stream()
-                        .noneMatch(excludeItem -> includeKeyPath.endsWith(excludeItem))).collect(Collectors.toList());
-    }
+    private void startProcessingObject(final int waitTimeMillis) {
+        final Optional<SourcePartition<S3SourceProgressState>> objectToProcess = sourceCoordinator.getNextPartition(this::provideFilteredBucketKeyPartitions);
 
-    private void processS3ObjectKeys(final S3ObjectReference s3ObjectReference,
-                                     final S3ObjectHandler s3ObjectHandler,
-                                     final ScanOptions scanOptions){
-        final S3ObjectDetails s3ObjDetails = getS3ObjectDetails(s3ObjectReference);
-        final boolean isKeyMatchedBetweenTimeRange = isKeyMatchedBetweenTimeRange(s3ObjDetails.getS3ObjectLastModifiedTimestamp(),
-                scanOptions.getUseStartDateTime(),
-                scanOptions.getUseEndDateTime());
-        if(isKeyMatchedBetweenTimeRange && (isKeyNotProcessedByS3Scan(s3ObjDetails))){
-            updateKeyProcessedByS3Scan(s3ObjDetails);
-            try{
-                s3ObjectHandler.parseS3Object(s3ObjectReference,null);
-            }catch (IOException ex){
-                deleteKeyProcessedByS3Scan(s3ObjDetails);
-                LOG.error("Error while process the parseS3Object. ",ex);
+        if (objectToProcess.isEmpty()) {
+            try {
```

Comment:
I think it is better to put this wait() in the calling function `run` instead of here.

Reply:
This structure is to make the testing easier. The alternative is to run the test with awaitility or an executor service, and to reflectively set the `shouldStopProcessing` boolean. If you feel strongly about it I can make this change.